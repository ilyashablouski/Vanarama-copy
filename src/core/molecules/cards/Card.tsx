import cx from 'classnames';
import React, { FC, memo } from 'react';

import BlackFridayCardLabel from 'core/molecules/cards/BlackFridayCardLabel';
import { ICardProps } from './interfaces';
import CardHeader from './CardHeader';
import CardTitle from './CardTitle';
import Text from '../../atoms/text';

import {
  optimiseImage,
  DEFAULT_OPTIMISATION,
} from '../../../helpers/imageOptimiseUtils';

const Card: FC<ICardProps> = memo(props => {
  const {
    className,
    alt,
    lazyLoad,
    children,
    inline,
    overflow,
    header,
    title,
    description,
    style,
    placeholderImage,
    optimisedHost,
    optimisationOptions,
    extrasRender,
    isBlackFridayFlag,
    dataUiTestId,
  } = props;

  const { imageSrc } = props;

  let srcset;
  let srcDefault = imageSrc;
  let src1200;
  let src800;
  let src320;

  // Check if image should be optimised.
  if (imageSrc !== undefined && optimisedHost) {
    srcDefault = optimiseImage(optimisedHost, imageSrc, {
      ...DEFAULT_OPTIMISATION,
      ...optimisationOptions,
    });
    src1200 = optimiseImage(optimisedHost, imageSrc, {
      ...DEFAULT_OPTIMISATION,
      width: 1200,
      height: 1200,
      ...optimisationOptions,
    });
    src800 = optimiseImage(optimisedHost, imageSrc, {
      ...DEFAULT_OPTIMISATION,
      width: 800,
      height: 800,
      quality: 70,
      ...optimisationOptions,
    });
    src320 = optimiseImage(optimisedHost, imageSrc, {
      ...DEFAULT_OPTIMISATION,
      width: 320,
      height: 320,
      quality: 60,
      ...optimisationOptions,
    });

    srcset = `${src320} 320w, ${src800} 800w, ${src1200} 1200w`;
  }

  const onImageError = (e: any) => {
    e.target.srcset = '';
    e.target.src = `${process.env.HOST_DOMAIN}/vehiclePlaceholder.jpg`;
  };

  return (
    <div
      className={cx('card', className, {
        '-inline': inline,
        '-overflow': overflow,
      })}
      data-testid="card"
      style={style}
      data-uitestid={dataUiTestId}
    >
      {header?.text && <CardHeader {...header} />}
      {imageSrc !== undefined ? (
        <>
          <img
            loading={lazyLoad ? 'lazy' : 'eager'}
            srcSet={srcset}
            sizes="(min-width:320px) 800px, 1200px"
            alt={alt}
            className="card-image"
            src={srcDefault || imageSrc || placeholderImage || undefined}
            data-cfasync="false"
            data-testid="card-image"
            onError={onImageError}
          />
          {isBlackFridayFlag && <BlackFridayCardLabel />}
        </>
      ) : (
        isBlackFridayFlag && <BlackFridayCardLabel className="-relative" />
      )}
      {extrasRender && <div className="extras">{extrasRender}</div>}
      {(title?.title || title?.link) && <CardTitle {...title} />}
      {description && (
        <Text color="dark" dataTestId="card-description">
          {description}
        </Text>
      )}
      {children}
    </div>
  );
});

export default Card;
