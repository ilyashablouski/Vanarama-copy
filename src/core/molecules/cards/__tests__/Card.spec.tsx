import React from 'react';
import { shallow } from 'enzyme';
import Card from '../Card';

describe('<Card />', () => {
  it('should render correctly with children', () => {
    // ACT
    const wrapper = shallow(
      <Card>
        <p>Hello World</p>
      </Card>,
    );

    // ASSERT
    expect(wrapper).toMatchSnapshot();
  });

  it('should render correctly with a custom class name', () => {
    // ACT
    const wrapper = shallow(
      <Card className="some-custom-class-name">
        <p>Hello World</p>
      </Card>,
    );

    // ASSERT
    expect(wrapper).toMatchSnapshot();
  });

  it('should render correctly only image', () => {
    // ACT
    const wrapper = shallow(
      <Card imageSrc="https://source.unsplash.com/collection/2102317/1000x650?sig=403440" />,
    );

    // ASSERT
    expect(wrapper).toMatchSnapshot();
  });

  it('Basic Card should render correctly', () => {
    // ACT
    const wrapper = shallow(
      <Card
        title={{
          title: 'Basic Card',
        }}
        description="Dolore Lorem culpa ea irure do ad exercitation cillum laborum culpa sint aliqua voluptate duis"
      />,
    );

    // ASSERT
    expect(wrapper).toMatchSnapshot();
  });

  it('Featured Review Card should render correctly', () => {
    // ACT
    const wrapper = shallow(
      <Card
        title={{
          title: '',
          link: (
            <a href="#" className="heading -large -black">
              Featured Review Card
            </a>
          ),
          score: 4.5,
        }}
        description="Minim consectetur adipisicing aute consequat velit exercitation enim deserunt occaecat sit ut incididunt dolor id"
        imageSrc="https://source.unsplash.com/collection/2102317/1000x650?sig=403440"
      />,
    );

    // ASSERT
    expect(wrapper).toMatchSnapshot();
  });

  it('Article Card should render correctly', () => {
    // ACT
    const wrapper = shallow(
      <Card
        title={{
          title: 'Article Card',
        }}
        description="Minim consectetur adipisicing aute consequat velit exercitation enim deserunt occaecat sit ut incididunt dolor id"
        imageSrc="https://source.unsplash.com/collection/2102317/1000x650?sig=403440"
      />,
    );

    // ASSERT
    expect(wrapper).toMatchSnapshot();
  });
});
