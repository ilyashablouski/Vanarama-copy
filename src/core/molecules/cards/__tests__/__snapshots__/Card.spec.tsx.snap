// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Card /> Article Card should render correctly 1`] = `
<div
  className="card"
  data-testid="card"
>
  <Memo(ImageV2)
    className="card-image"
    dataTestId="card-image"
    objectFit="cover"
    src="https://source.unsplash.com/collection/2102317/1000x650?sig=403440"
  />
  <Memo()
    title="Article Card"
  />
  <Memo(Text)
    color="dark"
    dataTestId="card-description"
  >
    Minim consectetur adipisicing aute consequat velit exercitation enim deserunt occaecat sit ut incididunt dolor id
  </Memo(Text)>
</div>
`;

exports[`<Card /> Basic Card should render correctly 1`] = `
<div
  className="card"
  data-testid="card"
>
  <Memo()
    title="Basic Card"
  />
  <Memo(Text)
    color="dark"
    dataTestId="card-description"
  >
    Dolore Lorem culpa ea irure do ad exercitation cillum laborum culpa sint aliqua voluptate duis
  </Memo(Text)>
</div>
`;

exports[`<Card /> Featured Review Card should render correctly 1`] = `
<div
  className="card"
  data-testid="card"
>
  <Memo(ImageV2)
    className="card-image"
    dataTestId="card-image"
    objectFit="cover"
    src="https://source.unsplash.com/collection/2102317/1000x650?sig=403440"
  />
  <Memo()
    link={
      <a
        className="heading -large -black"
        href="#"
      >
        Featured Review Card
      </a>
    }
    score={4.5}
    title=""
  />
  <Memo(Text)
    color="dark"
    dataTestId="card-description"
  >
    Minim consectetur adipisicing aute consequat velit exercitation enim deserunt occaecat sit ut incididunt dolor id
  </Memo(Text)>
</div>
`;

exports[`<Card /> should render correctly only image 1`] = `
<div
  className="card"
  data-testid="card"
>
  <Memo(ImageV2)
    className="card-image"
    dataTestId="card-image"
    objectFit="cover"
    src="https://source.unsplash.com/collection/2102317/1000x650?sig=403440"
  />
</div>
`;

exports[`<Card /> should render correctly with a custom class name 1`] = `
<div
  className="card some-custom-class-name"
  data-testid="card"
>
  <p>
    Hello World
  </p>
</div>
`;

exports[`<Card /> should render correctly with children 1`] = `
<div
  className="card"
  data-testid="card"
>
  <p>
    Hello World
  </p>
</div>
`;
